Separate general policy logic from block builder policy logic to improve usability for other TEE projects

spencer-ritual opened the pull request:
# Refactor: extract BasePolicy + deriver architecture; keep BlockBuilderPolicy upgrade-safe

## Summary
This PR refactors `BlockBuilderPolicy` by extracting reusable policy logic into a new abstract `BasePolicy`, and introduces a pluggable workload-derivation layer (`IWorkloadDeriver`) so workload ID derivation can evolve without rewriting policy code. The refactor is designed to preserve upgradeability by maintaining `BlockBuilderPolicy`’s storage layout.

## Previous Discussion
Discussion occurred in https://github.com/flashbots/flashtestations/issues/45#issuecomment-3478453628

## What changed
- New shared surface
  - Added `src/BasePolicy.sol` with common policy logic (`addWorkloadToPolicy`, `removeWorkloadFromPolicy`, `getWorkloadMetadata`, `isAllowedPolicy`, `_cachedIsAllowedPolicy`) and hooks for auth/deriver/cache
  - Added `src/interfaces/IPolicyCommon.sol` for shared `WorkloadId` type, `WorkloadMetadata`, and common events/errors
  - Added `src/interfaces/IBasePolicy.sol` for the shared policy interface
  - Added `src/interfaces/IWorkloadDeriver.sol` for injected workload derivation (`workloadIdForQuote(bytes)`)
- Deriver implementation
  - Added `src/derivers/TDXWorkloadDeriver.sol` containing the current TDX derivation logic behind `IWorkloadDeriver`
- Refactored `BlockBuilderPolicy`
  - `BlockBuilderPolicy` now inherits `BasePolicy` and keeps block-builder-specific logic (EIP-712 domain, proof verification, permit nonce handling)
  - Added configurable `workloadDeriver` and wired it through `initialize(...)` and `setWorkloadDeriver(...)`
  - Overrode `isAllowedPolicy` to derive from `registration.parsedReportBody` to avoid re-parsing raw quotes on cache misses
    - Added a runtime guard in `_setWorkloadDeriver` to ensure the configured deriver supports the report-body derivation method assumed by the override
- Examples
  - Added `examples/DualDeriverPolicy.sol` (UNAUDITED EXAMPLE) demonstrating a migration approach that tries an “old” and “new” deriver
  - Added `examples/TDXTD15WorkloadDeriver.sol` showing how a hypothetical future report-body format could be supported by swapping derivers
- Tests + scripts
  - Added `test/UpgradeRegression.t.sol` to upgrade a proxy from a legacy policy implementation to the refactored `BlockBuilderPolicy` and assert state + behavior preservation
  - Added/updated tests for the examples (`test/Examples.t.sol`) and updated existing tests/scripts for the new initializer/deriver wiring

## Why
- Reuse shared policy logic for other policy types beyond block-building
- Make workload ID derivation upgradeable/swappable via an injected deriver contract
- Preserve production upgrade paths by keeping `BlockBuilderPolicy` storage layout compatible

## Storage / upgradeability notes
- `BasePolicy` only introduces shared storage in the original order (`approvedWorkloads`, `registry`) and uses hooks so derived contracts can own additional state
- `BlockBuilderPolicy` keeps its existing slots and introduces `workloadDeriver` in reserved space; the upgrade regression test validates this end-to-end

## Test plan
- `forge test`

spencer-ritual committed:
draft

spencer-ritual committed:
modify comment to say that we enforce instead of assume

spencer-ritual committed:
make add and remove virtual

frieder-ritual commented:
Technically the refactor looks very good!

There are two things that potentially will be raised:

1. The repo is very document heavy, there are several places where the refactor has deleted comments all together or at least omitted details/reformulated with a slightly different meaning. I'll reference a version that partly restores comments and will also add a comment that diffs old and new comments where things have changed
2. I like that this PR is also addressing #49 and #52. But it makes the review process more difficult as even more things have moved.

frieder-ritual commented:
Below are the methods whose doc blocks changed meaningfully. For each: 1) old, 2) new

  **initialize**

  1. Old (IBlockBuilderPolicy)

  /// @notice Initializer to set the FlashtestationRegistry contract which verifies TEE quotes and the initial owner of the contract
  /// @param _initialOwner The address of the initial owner of the contract
  /// @param _registry The address of the registry contract

  2. New (IBlockBuilderPolicy)

  /// @notice Initializer to set the FlashtestationRegistry contract which verifies TEE quotes and the initial owner of the contract
  /// @param _initialOwner The address of the initial owner of the contract
  /// @param _registry The address of the registry contract
  /// @param _workloadDeriver Address of the workload deriver used for workloadId computation

  **isAllowedPolicy**

  1. Old (IBlockBuilderPolicy)

  /// @notice Check if this TEE-controlled address has registered a valid TEE workload with the registry, and
  /// if the workload is approved under this policy
  /// @param teeAddress The TEE-controlled address
  /// @return allowed True if the TEE is using an approved workload in the policy
  /// @return workloadId The workloadId of the TEE that is using an approved workload in the policy, or 0 if
  /// the TEE is not using an approved workload in the policy

  2. New (IBasePolicy + BasePolicy inline)

  /// @notice Check if this TEE-controlled address has a valid registry registration and
  /// whether its workload is approved under this policy.
  /// @param teeAddress The TEE-controlled address.
  /// @return allowed True if the TEE is using an approved workload in the policy.
  /// @return workloadId The workloadId of the TEE that is using an approved workload, or 0 if not allowed.

  **_cachedIsAllowedPolicy**

  1. Old (BlockBuilderPolicy)

  /// @notice isAllowedPolicy but with caching to reduce gas costs
  /// @dev This function is only used by the verifyBlockBuilderProof function, which needs to be as efficient as possible
  /// because it is called onchain for every flashblock. The workloadId is cached to avoid expensive recomputation
  /// @dev A careful reader will notice that this function does not delete stale cache entries. It overwrites them
  /// if the underlying TEE registration is still valid. But for stale cache entries in every other scenario, the
  /// cache entry persists indefinitely. This is because every other instance results in a return value of (false, 0)
  /// to the caller (which is always the verifyBlockBuilderProof function) and it immediately reverts. This is an unfortunate
  /// consequence of our need to make this function as gas-efficient as possible, otherwise we would try to cleanup
  /// stale cache entries
  /// @param teeAddress The TEE-controlled address
  /// @return True if the TEE is using an approved workload in the policy
  /// @return The workloadId of the TEE that is using an approved workload in the policy, or 0 if
  /// the TEE is not using an approved workload in the policy

  2. New (BasePolicy)

  /// @notice Cached variant of `isAllowedPolicy` for gas-sensitive call paths.
  /// @dev Why this exists:
  ///      - `isAllowedPolicy` is `view` but can be expensive because workloadId derivation can be costly.
  ///      - Some policies (notably `BlockBuilderPolicy`) need an O(1)-ish authorization check on the hot path
  ///        (e.g. `verifyBlockBuilderProof`), so we cache the derived workloadId keyed by the TEE's quoteHash.
  ///
  /// @dev How it is expected to be used:
  ///      - Derived policies call `_cachedIsAllowedPolicy(teeAddress)` from their hot path.
  ///      - Cache storage lives in the derived policy; `BasePolicy` only implements the algorithm and calls the
  ///        `_getCachedWorkload` / `_setCachedWorkload` hooks.
  ///      - The function checks registry validity + quoteHash; on cache hit it avoids recomputing derivation.
  ///      - On cache miss (or quote change), it falls back to `isAllowedPolicy` and updates the cache if allowed.
  ///
  /// @dev Cache cleanup:
  ///      - This function does not proactively delete stale cache entries. On invalid registrations it returns
  ///        `(false, 0)` and callers typically revert, so cleanup is intentionally skipped to keep the hot path cheap.
  ///      - Stale cache entries for all other "not allowed" scenarios persist indefinitely for the same reason:
  ///        the caller reverts immediately and the hot path avoids extra storage writes.
  ///
  /// @param teeAddress The TEE-controlled address to check.
  /// @return allowed True if the TEE is registered, valid, and running an approved workload.
  /// @return workloadId The approved workloadId for this TEE (or 0 if not allowed).

  **workloadIdForTDRegistration**

  1. Old (IBlockBuilderPolicy + inline)

  /// @notice Application specific mapping of registration data to a workload identifier
  /// @dev Think of the workload identifier as the version of the application for governance.
  /// The workloadId verifiably maps to a version of source code that builds the TEE VM image
  /// @param registration The registration data from a TEE device
  /// @return workloadId The computed workload identifier

  Inline:

  // We expect FPU and SSE xfam bits to be set, and anything else should be handled by explicitly allowing the workloadid
  // We don't mind VE_DISABLED, PKS, and KL tdattributes bits being set either way, anything else requires explicitly allowing the workloadid

  2. New (IBlockBuilderPolicy + BlockBuilderPolicy)

  /// @notice Application specific mapping of registration data to a workload identifier
  /// @dev Think of the workload identifier as the version of the application for governance.
  /// The workloadId verifiably maps to a version of source code that builds the TEE VM image
  /// @param registration The registration data from a TEE device
  /// @return workloadId The computed workload identifier

  Inline:

  // Uses TDXWorkloadDeriverLib (see the deriver for constants and derivation details).

  **addWorkloadToPolicy**

  1. Old (IBlockBuilderPolicy)

  /// @notice Add a workload to a policy (governance only)
  /// @notice Only the owner of this contract can add workloads to the policy
  /// and it is the responsibility of the owner to ensure that the workload is valid
  /// otherwise the address associated with this workload has full power to do anything
  /// who's authorization is based on this policy
  /// @dev The commitHash solves the following problem; ...
  /// @param workloadId The workload identifier
  /// @param commitHash The 40-character hexadecimal commit hash of the git repository
  /// whose source code is used to build the TEE image identified by the workloadId
  /// @param sourceLocators An array of URIs pointing to the source code

  2. New (IBasePolicy)

  /// @notice Add a workload to a policy (governance only).
  /// @notice Only the policy authority can add workloads to the policy, and it is the responsibility of the
  /// authority to ensure that the workload is valid; otherwise the address associated with this workload has
  /// full power to do anything whose authorization is based on this policy.
  /// @dev The commitHash solves the following problem; ...
  /// @param workloadId The workload identifier.
  /// @param commitHash The 40-character hexadecimal commit hash ...
  /// @param sourceLocators An array of URIs pointing to the source code.


  **removeWorkloadFromPolicy**

  1. Old (IBlockBuilderPolicy)

  /// @notice Remove a workload from a policy (governance only)
  /// @param workloadId The workload identifier

  2. New (IBasePolicy)

  /// @notice Remove a workload from a policy (governance only).

  **getWorkloadMetadata**

  1. Old (IBlockBuilderPolicy)

  /// @notice Mapping from workloadId to its metadata (commit hash and source locators)
  /// @dev This is only updateable by governance (i.e. the owner) of the Policy contract
  /// Adding and removing a workload is O(1)
  /// @param workloadId The workload identifier to query
  /// @return The metadata associated with the workload

  2. New (IBasePolicy)

  /// @notice Get metadata for an approved workload.
  /// @dev This is only updateable by governance (i.e. the policy authority). Adding and removing a workload is O(1).
  /// @param workloadId The workload identifier to query.
  /// @return The metadata associated with the workload.

frieder-ritual committed:
Restore policy docs after refactor

frieder-ritual commented:
https://github.com/flashbots/flashtestations/pull/56

spencer-ritual commented:
Thank you for the feedback Frieder. I merged in your doc changes.

Regarding the other issues this PR addresses, this was a happy accident. Due to the requirement that the storage layout remain unchanged, BasePolicy cannot inherit from OwnableUpgradeable to access the owner state directly. Instead, it uses an abstract hook pattern (_checkPolicyAuthority()) that BlockBuilderPolicy implements by calling _checkOwner(). Similarly, before the tdx policy verification logic was BlockBuilderPolicy, but we decided to make derivers that needed this logic, and I wouldn't want a deriver to inherit from BlockBuilderPolicy so it just made sense to refactor out. 

@Melvillian do you think you could take a look and let me know what you think?

Melvillian commented on src/BlockBuilderPolicy.sol:
```diff
@@ -47,60 +42,83 @@ contract BlockBuilderPolicy is
     /// @dev Maps teeAddress to cached workload information for gas optimization
     mapping(address teeAddress => CachedWorkload) private cachedWorkloads;
 
+    /// @notice Workload deriver used by the shared base policy logic (Option B).
```
readability: Option B only has meaning to us devs who have worked on this particular PR, but it won't mean anything to most future readers of this code. Please replace "(Option B)" within a concise explanation of why we're going with the WorkloadDeriver approach, so it's understandable by everyone

spencer-ritual replied:
addressed in [ed68a40](https://github.com/flashbots/flashtestations/pull/55/commits/ed68a40f7c3e24e2c3314776464a716c8e8614e3)

Melvillian commented on src/BlockBuilderPolicy.sol:
```diff
@@ -47,60 +42,83 @@ contract BlockBuilderPolicy is
+    IWorkloadDeriver public workloadDeriver;
+
+    /// @notice Emitted when the workload deriver is set or updated.
+    event WorkloadDeriverSet(address indexed deriver);
```
code-structure-nit: Should this go in `IWorkloadDeriver.sol`? Seems like it should, since we're putting other `event`s in interfaces like `IPolicyCommon.sol`

spencer-ritual replied:
addressed in [e5930b7](https://github.com/flashbots/flashtestations/pull/55/commits/e5930b76e524141f2dc6c505d72ac0c67555266f)

Melvillian commented on src/BasePolicy.sol:
```diff
@@ -0,0 +1,193 @@
+    /// @param teeAddress The TEE-controlled address to check.
+    /// @return allowed True if the TEE is registered, valid, and running an approved workload.
+    /// @return workloadId The approved workloadId for this TEE (or 0 if not allowed).
+    function _cachedIsAllowedPolicy(address teeAddress) internal returns (bool, WorkloadId) {
```
Note: just for your awareness, I think this should be made `public` (see https://github.com/flashbots/flashtestations/issues/57). We don't have to do it here because this PR is already so large, but I'm making you aware of it.

Melvillian commented:
> Technically the refactor looks very good!
> 
> There are two things that potentially will be raised:
> 
> 1. The repo is very document heavy, there are several places where the refactor has deleted comments all together or at least omitted details/reformulated with a slightly different meaning. I'll reference a version that partly restores comments and will also add a comment that diffs old and new comments where things have changed
> 2. I like that this PR is also addressing [Separate roles for contract owner and workload ID permissioner #49](https://github.com/flashbots/flashtestations/issues/49) and [Abstract the TEE policy verification logic into a library useable by external contracts #52](https://github.com/flashbots/flashtestations/issues/52). But it makes the review process more difficult as even more things have moved.

Note that this PR doesn't implement #49 fully, though it does lay some necessary groundwork in that it adds the `setWorkloadDeriver` function. This PR still uses the same `owner` variable for upgrading the contract as it does for calling `setWorkloadDeriver`.

In order to fully implement #49, we need to refactor the ownership logic into two separate "roles", for which I would use OZ's [AccessControl contract](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/AccessControl.sol).

I think the switch to AccessControl would be a large change, and this PR is already so large so we shouldn't do it here. But be sure not to update #49 as closed when this PR merges, because that wouldn't be correct.

Melvillian commented on src/BlockBuilderPolicy.sol:
```diff
@@ -33,6 +27,7 @@ struct CachedWorkload {
  * is allowed under any workload in a Policy, and the FlashtestationRegistry will handle the rest
  */
 contract BlockBuilderPolicy is
+    BasePolicy,
```
This ordering of parent contract is wrong because it will not give us the backwards-compatible storage variable ordering we need, i.e. the same ordering as `LegacyBlockBuilderPolicy`. The ordering should have `BasePolicy` be last, because that means `BasePolicy`'s constructor will be executed right before `BlockBuilderPolicy`'s constructor is executed.

Please double-check my reasoning here. I am looking at [the multiple inheritance solidity docs here](https://docs.soliditylang.org/en/v0.8.33/contracts.html#multiple-inheritance-and-linearization), which implies that constructors will be run in the order they are listed in the contract's inheritance definition.

Another way to understand what I'm saying is that because you have `BasePolicy` written as the first parent contract, then its storage variables will be first in the final `BlockBuilderPolicy`, then you'll have `Initializable`'s, then `UUPSUpgradeable`'s, then `OwnableUpgradeable`'s, and finally `BlockBuilderPolicy`'s. But that doesn't match the `LegacyBlockBuilderPolicy`'s, which has all of the OZ contract's constructors run first, and then finally it runs its `LegacyBlockBuilderPolicy` constructor which will _first_ write the equivalent of the `BasePolicy`'s storage variables, followed finally by its own storage variables.

Melvillian commented on src/derivers/TDXWorkloadDeriver.sol:
```diff
@@ -0,0 +1,70 @@
+
+    /// @dev See section 11.5.3 in TDX Module v1.5 Base Architecture Specification.
+    /// @notice Enabled FPU (always enabled).
+    bytes8 internal constant TD_XFAM_FPU = 0x0000000000000001;
```
Note: you're using a stale version of the workload derivation logic. We removed the TDX workload constants logic in https://github.com/flashbots/flashtestations/pull/53 because we found a bug in it, and rather than try to fix it and maintain workload constant flexibility, we sacrificed workload constant flexibility in return for greater code simplicity and thus security.

See the latest BlockBuilderPolicy.sol for what the workload derivation code should look like:

https://github.com/flashbots/flashtestations/blob/main/src/BlockBuilderPolicy.sol

spencer-ritual replied:
addressed in [daadeef](https://github.com/flashbots/flashtestations/pull/55/commits/daadeef14ce4da2e900d6c7e60a7934ae5ccabb7)

Melvillian commented on examples/TDXTD15WorkloadDeriver.sol:
```diff
@@ -0,0 +1,123 @@
+    using BytesUtils for bytes;
+
+    // Same constants as the current TDX deriver.
+    bytes8 internal constant TD_XFAM_FPU = 0x0000000000000001;
```
Note: you're using a stale version of the workload derivation logic. We removed the TDX workload constants logic in #53 because we found a bug in it, and rather than try to fix it and maintain workload constant flexibility, we sacrificed workload constant flexibility in return for greater code simplicity and thus security.

See the latest BlockBuilderPolicy.sol for what the workload derivation code should look like:

https://github.com/flashbots/flashtestations/blob/main/src/BlockBuilderPolicy.sol

spencer-ritual replied:
addressed in [daadeef](https://github.com/flashbots/flashtestations/pull/55/commits/daadeef14ce4da2e900d6c7e60a7934ae5ccabb7)

Melvillian commented on test/fixtures/LegacyBlockBuilderPolicy.sol:
```diff
@@ -0,0 +1,188 @@
+        keccak256("VerifyBlockBuilderProof(uint8 version,bytes32 blockContentHash,uint256 nonce)");
+
+    // TDX workload constants (legacy)
+    bytes8 constant TD_XFAM_FPU = 0x0000000000000001;
```
Note: you're using a stale version of the workload derivation logic. We removed the TDX workload constants logic in #53 because we found a bug in it, and rather than try to fix it and maintain workload constant flexibility, we sacrificed workload constant flexibility in return for greater code simplicity and thus security.

See the latest BlockBuilderPolicy.sol for what the workload derivation code should look like:

https://github.com/flashbots/flashtestations/blob/main/src/BlockBuilderPolicy.sol

spencer-ritual replied:
addressed in [daadeef](https://github.com/flashbots/flashtestations/pull/55/commits/daadeef14ce4da2e900d6c7e60a7934ae5ccabb7)

Melvillian commented on test/UpgradeRegression.t.sol:
```diff
@@ -0,0 +1,117 @@
+import {Upgrader} from "./helpers/Upgrader.sol";
+import {Helper} from "./helpers/Helper.sol";
+
+contract UpgradeRegressionTest is Test {
```
note: i double-checked the that the storage layouts of LegacyBlockBuilderPolicy and BlockBuilderPolicy match using `forge inspect <Contract Name> storage-layout --pretty` and can confirm they are ordered as expected! Nice job.

Melvillian reviewed (changes requested):
I looked through everything and added some inline request for changes. These are blocking, but they are a small number of changes and overall I think this looks solid. Thank you for putting in the work to abstract out these difference modules, it's a real improvement.

Note, I did not do an audit-level review of the all the tests, so I don't have high confidence this is secure. But I did look through every contract and pointed out parts I'm pretty sure need to be modified.

spencer-ritual committed:
docs: clarify workload deriver composition pattern

Remove "(Option B)" reference and add explanation for why the deriver uses
a composition pattern, making the design rationale clearer for future maintainers.

spencer-ritual committed:
refactor: move WorkloadDeriverSet event to interface

Move event declaration from BlockBuilderPolicy implementation to
IBlockBuilderPolicy interface, following the pattern of other events in the
codebase and maintaining interface as the source of truth for public API.

spencer-ritual committed:
fix: update workload derivation to simplified logic from PR #53

Remove buggy TDX workload constants and masking logic. The new approach
directly hashes xFAM and tdAttributes without XOR/AND operations, trading
workload constant flexibility for code simplicity and security.

Updated files:
- src/derivers/TDXWorkloadDeriver.sol
- examples/TDXTD15WorkloadDeriver.sol
- test/fixtures/LegacyBlockBuilderPolicy.sol

spencer-ritual committed:
Remove test for tdAttributes bit masking behavior

The test_workloadId_tdAttributes_allowed_bits_ignored test verified that certain tdAttributes bits were ignored in workloadId computation. This behavior is no longer expected, so the test has been removed.

spencer-ritual commented:
Hey @Melvillian, I addressed the blocking changes if you could take a second look

spencer-ritual commented:
Hey @Melvillian, wanted to check in on this again

Melvillian commented on src/BlockBuilderPolicy.sol:
```diff
@@ -47,60 +42,81 @@ contract BlockBuilderPolicy is
+    }
+
+    function _setWorkloadDeriver(address deriver) internal {
+        require(deriver != address(0), "InvalidWorkloadDeriver");
```
style-requirement: we use custom errors throughout the contract, so we should continue to do so here. Please make these custom errors

Melvillian commented on src/BlockBuilderPolicy.sol:
```diff
@@ -47,60 +42,81 @@ contract BlockBuilderPolicy is
+
+    function _setWorkloadDeriver(address deriver) internal {
+        require(deriver != address(0), "InvalidWorkloadDeriver");
+        require(deriver.code.length > 0, "InvalidWorkloadDeriver");
```
style-requirement: we use custom errors throughout the contract, so we should continue to do so here. Please make these custom errors

Melvillian commented on src/BlockBuilderPolicy.sol:
```diff
@@ -47,60 +42,81 @@ contract BlockBuilderPolicy is
+        TD10ReportBody memory empty;
+        (bool ok, bytes memory ret) =
+            deriver.staticcall(abi.encodeCall(TDXWorkloadDeriver.workloadIdForReportBody, (empty)));
+        require(ok && ret.length == 32, "DeriverMissingWorkloadIdForReportBody");
```
style-requirement: we use custom errors throughout the contract, so we should continue to do so here. Please make these custom errors

Melvillian reviewed (changes requested):
Looks good. See the comments about custom errors, once those are fixed I'll approve

spencer-ritual committed:
refactor: replace string error messages with custom errors in BlockBuilderPolicy

Replace string-based require error messages with custom errors in `_setWorkloadDeriver`
to maintain consistency with the contract's error handling pattern. Adds `InvalidWorkloadDeriver`
and `DeriverMissingWorkloadIdForReportBody` custom errors to the interface.

spencer-ritual commented:
Hey, the three places you flagged should be addressed
